core/src/main/java/com/linecorp/armeria/common/stream/DefaultStreamMessage.java

```diff
@@ -305,25 +316,49 @@ private boolean notifySubscriber(Subscriber<Object> subscriber, Queue<Object> qu
 
             if (demand == Long.MAX_VALUE || demandUpdater.compareAndSet(this, demand, demand - 1)) {
                 @SuppressWarnings("unchecked")
-                final T o = (T) queue.remove();
+                T o = (T) queue.remove();
+                if (o instanceof ReferenceCounted) {
+                    ((ReferenceCounted) o).touch();
+                }
+
                 onRemoval(o);
-                if (!subscription.withPooledObjects() && o instanceof ByteBufHttpData) {
-                    ByteBuf buf = ((ByteBufHttpData) o).buf();
-                    try {
-                        subscriber.onNext(HttpData.of(ByteBufUtil.getBytes(buf)));
-                    } finally {
-                        buf.release();
+                if (!subscription.withPooledObjects()) {
+                    if (o instanceof ByteBufHolder) {
+                        o = copyAndRelease((ByteBufHolder) o);
+                    } else if (o instanceof ByteBuf) {
+                        o = copyAndRelease((ByteBuf) o);
                     }
-                } else {
-                    subscriber.onNext(o);
                 }
+
+                subscriber.onNext(o);
                 return true;
             }
         }
 
         return false;
     }
 
+    private T copyAndRelease(ByteBufHolder o) {
+        try {
+            final ByteBuf content = Unpooled.wrappedBuffer(ByteBufUtil.getBytes(o.content()));
+            @SuppressWarnings("unchecked")
+            final T copy = (T) o.replace(content);
+            return copy;
+        } finally {
+            ReferenceCountUtil.safeRelease(o);
```

Comment:
Since we use safeRelease either way, I guess the methods could just copy and the caller can release. to avoid the duplication.